contracts/cw-storage/src/contract.rs

```diff
@@ -116,6 +120,78 @@ pub mod execute {
 
         Ok(res)
     }
+
+    pub fn pin_object(
+        deps: DepsMut,
+        info: MessageInfo,
+        object_id: ObjectId,
+    ) -> Result<Response, ContractError> {
+        if pins().has(deps.storage, (object_id.clone(), info.sender.clone())) {
+            return Ok(Response::new()
+                .add_attribute("action", "pin_object")
+                .add_attribute("id", object_id));
+        }
+
+        let bucket = BUCKET.load(deps.storage)?;
+
+        let o = objects().update(
+            deps.storage,
+            object_id.clone(),
+            |o| -> Result<Object, StdError> {
+                o.map(|mut e: Object| -> Object {
+                    e.pin_count += Uint128::one();
+                    e
+                })
+                .ok_or_else(|| StdError::not_found(type_name::<Object>()))
+            },
+        )?;
```

Comment:
We may load the bucket after updating the object, it'll reduce the gas consumption in case the object doesn't exists.